searches/knuth_morris_pratt_algorithm.py

```diff
@@ -0,0 +1,73 @@
+# https://en.wikipedia.org/wiki/Knuth%E2%80%93Morris%E2%80%93Pratt_algorithm
+
+
+def knuth_morris_pratt(text: str, pattern: str) -> list[int]:
```

Comment:
There are a couple edgecases you could add support for:
- an empty pattern
- a pattern longer than the text